### ISSUE TITLE
Should apply level hierarchy

### ISSUE THREAD

**Issue by ecureuill** [solex2006/SELIProject#64]
## Expected Behavior
Programmatic heading levels should match their visual appearance (like size and boldness).

## Current Behavior

Text contents of type Title, create by teacher, are not rendered using the correct heading tags. Big, Medium and Normal title are all coded as `h2`
![image](https://user-images.githubusercontent.com/993369/78851915-68c45d00-79f1-11ea-8165-ec6c70c14ae9.png)

## Steps to Reproduce

1. As a teacher, create a course with text contents
   1. Add a Title content of size **Big**
   1. Add a Title content of size **Medium**
   1. Add a Title content of size **Normal**
1. As a student, navigate to the course and access it content
   1. Check the html of each of title contents

## How to fix
Structure multiple headings on a page hierarchically. (best practice)
For example, try to follow nested content under an h2 heading with h3 before you use h4.
Exception: For fixed content that repeats across pages (like a footer or a sidebar), the heading level should not change.

Follow #62 and #63 specification

## Related issues
#64

**Comment by juantenesaca** [comment]
Hi Camilla, So could you provide as a table specifying which hierarchy should have each component. Then, we will not have more changes for it.

**Comment by ecureuill** [comment]
@juantenesaca 

The hierarchy is the html specification `h1` > `h2` > `h3` > `h4` >` h5` > `h6`

![image](https://user-images.githubusercontent.com/993369/80423514-8715ce00-88b6-11ea-9bd9-93d0b1769087.png)

[Examples](https://www.nomensa.com/blog/2017/how-structure-headings-web-accessibility)

As the Authoring Tool provides 3 titles level  (sections) to teacher, thus any **Section** content created inside one of those 3 titles level should follow the hierarchy (be a child). It's not a "fixed" solution. 

I will create some scenarios, but this list is not exhaustive, you will need to avail the html for each component to check if need any adjustment.

Note 1: as stated in #63 specification, a page must have just ONE top-level, thus i'm keeping `h1` reserved to be used for topic/unit name (student view). You need to check with the dev who is working with student view and instructional features, if he is planing to use any other sections, if this is the case, instead of start in `h2` as I'm proposing bellow, you should start on the next hierarchy level.

Component name | header
---|--- 
Text component - Big Title  | `<h2>`
Text component - Medium Title  |` <h3>`
Text component - Normal Title  | `<h4>`

Give above configuration,

Scenario: A component inside **Big Title** Section
For any component that is composed by a section, the top-level header must start at `<h3>`
Other sections levels, must follows hierarchy

Scenario: A component inside **Medium Title** Section
For any component that is composed by a section, the top-level header must start at `<h4>`
Other sections levels, must follows hierarchy

Scenario: A component inside **Normal Title** Section
For any component that is composed by a section, the top-level header must start at `<h5>`
Other sections levels, must follows hierarchy

Scenario: A component **out of any Title Section**
For any component that is composed by a section, the top-level header must start at `<h2>`
Other sections levels, must follows hierarchy

Note 2 : In examples bellow I'm using Activity and Quiz components, but this should be generalized for any component composed by sections, as stated in Scenarios.  
****************

<samp>
<h1><code>h1</code>Data Structure[Topic Name - inserted by system]</h1>
some texts
some text
<h2><code>h2</code> Activity one</h2>
Some text
<h2><code>h2</code> Quiz one</h2>
Some question
</samp>

****************

<samp>
<h1><code>h1</code>Data Structure[Topic Name - inserted by system]</h1>
<h2><code>h2</code>Data types [BIG TITLE component]</h2>
some text
<h2><code>h2</code>Basic data structures [BIG TITLE component]</h2>
some text
some video
<h3><code>h3</code> Activities [MIDDLE TITLE component]</h3>
A list of tasks
<h4><code>h4</code> Activity one</h4>
Some text
<h4><code>h4</code> Quiz one</h4>
Some question
</samp>
<h3><code>h3</code>List [MIDDLE TITLE component]</h3>
Some question
<h4><code>h4</code>Linked-List [NORMAL TITLE component]</h4>
Some question
<h5><code>h5</code> Activity one</h5>
Some text

**Comment by juantenesaca** [comment]
Thank you for the information. So, we have twelve items:

![image](https://user-images.githubusercontent.com/44218086/80429690-41ef9d00-88b2-11ea-95dd-3c91ccfe6308.png)

So, could you specify which hierarchy should have each component (then no more changes will be done). And we can close intermediately this issue, due the short time for developing.

**Comment by ecureuill** [comment]
@juantenesaca  it's not the case of a hierarchy for each component. The hierarchy will de determined by the presence of the **BIG[MEDIUM|NORMAL title** (Texto Component configured to be titles) as parent of a content.

First, lets make some definition

**Target Components** are the authoring tool components that are builded (html) using headers tag
```html
<div><-- Some hypothetical component
  This component  is a TARGET COMPONENT
  Because there is one or more header tag in it -->
   <h4>
       <button>...</button>
   </h4>
   <div>
      <h5>...</h5> 
       <p>...</p>
    </div>  
</div>
```
```html
<div>
  <-- Another hypothetical component
  This component is not a TARGET COMPONENT
  Because there is no header tag in it -->
  <button>...</button>
   <div>
      <p>...</p>
    </div>  
</div>
```

Now, rewriting the scenarios I've stated in previous commentary 
When teacher add a **Target Component** in the _work area_, system must fix the headers tag
- If teacher add a **Target Component** as child of a **BIG Title**, thus system must adjust the **Target Component**'s headers to start in `h3` 
Using the above hypothetical TARGET component, the final result will be
```html
<div>
   <h3><-- changed h4 by h3-->
       <button>...</button>
    </h4>
   <div>
      <h5>...</h5> <-- changed h5 by h4-->
      <p>...</p>
    </div>  
</div>
```
- If teacher add a **Target Component** as child of a **MEDIUM Title**, thus system must adjust the **Target Component**'s headers to start in `h4` 
Using the above hypothetical TARGET component, no change is necessary
```html
<div><-- Some hypothetical component
  This component  is a TARGET COMPONENT
  Because there is one or more header tag in it -->
   <h4>
       <button>...</button>
   </h4>
   <div>
      <h5>...</h5> 
       <p>...</p>
    </div>  
</div>
```
- If teacher add a **Target Component** as child of a **NORMAL Title**, thus system must adjust the **Target Component**'s headers to start in `h5` 
Using the above hypothetical TARGET component, the final result will be
```html
<div>
   <h5><-- changed h4 by h5-->
       <button>...</button>
    </h5>
   <div>
      <h6>...</h6> <-- changed h5 by h6-->
      <p>...</p>
    </div>  
</div>
```
- If teacher add a **Target Component** as NOT a child of any  **Title**, thus system must adjust the **Target Component**'s headers to start in `h2` 
Using the above hypothetical TARGET component, the final result will be
```html
<div>
   <h2><-- changed h4 by h2-->
       <button>...</button>
    </h2>
   <div>
      <h3>...</h3> <-- changed h5 by h3-->
      <p>...</p>
    </div>  
</div>
```
- [ ] So, you need to create an algorithm that will validate the Title hierarchy ( h2 > h3 > h4 > h5 > h6) of the contents added to the work area. This algorithm will just run for TARGET COMPONENTS.
- You need to determine which authoring tools are **target components** by looking for headers 
in theirs code
- As I said in the previous commentary, this i not an exhaustive list of scenarios. I listed the scenarios triggered by the "adding a target component in work area", but the validation need to considere other triggers event, like:
   - **Create** a BIG | MEDIUM | NORMAL Title as a parent of components already added in work area
   - **Move** a BIG | MEDIUM | NORMAL Title in work area
   - **Delete** a BIG | MEDIUM | NORMAL Title from work area
  -  **Move** a target component

************
#### Example of adding a hypothetical target component in work area, with no title as parent
![image](https://user-images.githubusercontent.com/993369/80443806-741af200-88e6-11ea-8a79-18ed27be05e2.png)

#### Example of adding a hypothetical target component in work area with a BIG TITLE as parent
![image](https://user-images.githubusercontent.com/993369/80443899-ae848f00-88e6-11ea-95df-197275556a93.png)

#### Example of adding a hypothetical target component in work area with a MEDIUM TITLE as parent
![image](https://user-images.githubusercontent.com/993369/80446911-68332e00-88ee-11ea-8567-4756f797a9fb.png)

#### Example of adding a hypothetical target component in work area with a NORMAL TITLE as parent
![image](https://user-images.githubusercontent.com/993369/80446991-9dd81700-88ee-11ea-80d4-1d9ecd1c7697.png)

#### Example of MOVING a Title that will cause the need to adust header of an existent Target Content 
- Inital state
![image](https://user-images.githubusercontent.com/993369/80447149-feffea80-88ee-11ea-9298-af93fd53d259.png)
- moving Title
![image](https://user-images.githubusercontent.com/993369/80447123-f0193800-88ee-11ea-9f1b-015a8a5f8a6e.png)
- final state
![image](https://user-images.githubusercontent.com/993369/80447039-bba57c00-88ee-11ea-957f-0b600309dc15.png)

#### Example of DELETING a Title that will cause the need to adust header of an existent Target Content 
- Inital state
![image](https://user-images.githubusercontent.com/993369/80447149-feffea80-88ee-11ea-9298-af93fd53d259.png)
- Deleting
![image](https://user-images.githubusercontent.com/993369/80447243-48503a00-88ef-11ea-83be-8d4535e533b0.png)
- final state
![image](https://user-images.githubusercontent.com/993369/80447274-5c943700-88ef-11ea-9570-2ac39eee56fd.png)

**Comment by juantenesaca** [comment]
I see. For it then, we need to first implement code to detect children and parents, in order to configure on which hierarchy they have to start.

@gavixo we need to implement more code for this requirement. What do you think?

**Comment by ecureuill** [comment]
@juantenesaca @gavixo have a look in branch [feature/64](https://github.com/solex2006/SELIProject/tree/feature/64)

I've done a header validation for my text Editor (a11yEditor) component that find the headers with wrong level and adjust existent headers new headers are added or when a existent header is deleted or had level changed.

I commited a simple version of that algorithm with a Demo component for test purpose

![image](https://user-images.githubusercontent.com/993369/80625841-23111800-8a24-11ea-9e0c-d7818094d651.png)

*************

What you think I create a new feature about the headers validation, because this is bigger scope then this bug ?  

And for this bug issue you just need to correct the header tags for:
Big Title -- `h2` 
Medium -- `h3`
Normal -- `h4`
 
Note: I'm suggesting start in `h2` supposing that only `h1` will be reserved fot the Topic/Unit page in Student View. You need to heck this with Instructional Developer

**Comment by ecureuill** [comment]
@juantenesaca, as I commented before, it's necessary a header validation and that the headers used in components be settled dynamically in Course's Page take in account the Title structured created by teacher.

:bulb:  My suggestion for now is that **you can avoid a lot of work by removing unnecessarily headers from components**

- Activity and Quiz: is coded using accordion (hilde/show panel) and for that header is required, thus I suggest that you render the activity and quiz without accordion

- Audio, Image and Video: follow their respective specifications in Instructional Design milestone that you will avoid the use of unnecessarily headers

Let's talk better when you start this.
